piksi_tools/console/tracking_view.py

```diff
@@ -133,104 +133,123 @@ class TrackingView(CodeFiltered):
+                    cno_array.append(0)
+            self.clean_cn0(t)
+            if self.update_scheduled:
+                return
+            self.update_scheduled = True
         GUI.invoke_later(self.update_plot)
 
     def tracking_state_callback(self, sbp_msg, **metadata):
-        codes_that_came = []
-        self.CN0_lock.acquire()
-        t = time.time() - self.t_init
-        self.time.append(t)
-        # first we loop over all the SIDs / channel keys we have stored and set 0 in for CN0
-        # for each SID, an array of size MAX PLOT with the history of CN0's stored
-        # If there is no CN0 or not tracking for an epoch, 0 will be used
-        # each array can be plotted against host_time, t
-        for i, s in enumerate(sbp_msg.states):
-            if code_is_glo(s.sid.code):
-                sat = s.fcn - GLO_FCN_OFFSET
-                self.glo_slot_dict[i] = s.sid.sat
-            else:
-                sat = s.sid.sat
-            key = (s.sid.code, sat, i)
-            codes_that_came.append(key)
-            if s.cn0 != 0:
-                self.CN0_dict[key].append(s.cn0 / 4.0)
-            received_code_list = getattr(self, "received_codes", [])
-            if s.sid.code not in received_code_list:
-                received_code_list.append(s.sid.code)
-                self.received_codes = received_code_list
-        for key, cno_array in self.CN0_dict.items():
-            if key not in codes_that_came:
-                cno_array.append(0)
-        self.CN0_lock.release()
+        with self.CN0_lock:
+            codes_that_came = []
+            t = time.time() - self.t_init
+            self.time.append(t)
+            # first we loop over all the SIDs / channel keys we have stored and set 0 in for CN0
+            # for each SID, an array of size MAX PLOT with the history of CN0's stored
+            # If there is no CN0 or not tracking for an epoch, 0 will be used
+            # each array can be plotted against host_time, t
+            for i, s in enumerate(sbp_msg.states):
+                if code_is_glo(s.sid.code):
+                    sat = s.fcn - GLO_FCN_OFFSET
+                    self.glo_slot_dict[i] = s.sid.sat
+                else:
+                    sat = s.sid.sat
+                key = (s.sid.code, sat, i)
+                codes_that_came.append(key)
+                if s.cn0 != 0:
+                    self.CN0_dict[key].append(s.cn0 / 4.0)
+                    self.CN0_age[key] = t
+                received_code_list = getattr(self, "received_codes", [])
+                if s.sid.code not in received_code_list:
+                    received_code_list.append(s.sid.code)
+                    self.received_codes = received_code_list
+            for key, cno_array in list(self.CN0_dict.items()):
+                if key not in codes_that_came:
+                    cno_array.append(0)
+            self.clean_cn0(t)
+            if self.update_scheduled:
```

Comment:
Avoid scheduling an update if one is already in-flight, this should reduce some of the processing overhead for plotting.